### PR TITLE
feat(binaries): improve platform and architecture support

### PR DESCRIPTION
**Pull request by HKGx**
## Change Summary

Resolves properly #195 
Pretty much just porting https://github.com/prisma/engines-wrapper to Python.
## Checklist

- [x] Unit tests for the changes exist
- [x] Tests pass without significant drop in coverage
- [x] Documentation reflects changes where applicable
- [x] Test snapshots have been [updated](https://prisma-client-py.readthedocs.io/en/latest/contributing/contributing/#snapshot-tests) if applicable

## Agreement

By submitting this pull request, I confirm that you can use, modify, copy and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by RobertCraigie** [comment]
Thank you! This is a much needed improvement.

Could you please add unit tests to ensure that all possible URLs that we generate will not result in a 404. You could base it off of this script (but should probably parallelise the requests with the `AsyncClient`):

```py
import httpx
from prisma.binaries import ENGINES

url = ENGINES[0].url
print(url)

client = httpx.Client()
response = client.head(url)
response.raise_for_status()

client.close()
```

**Comment by HKGx** [comment]
> Could you please add unit tests to ensure that all possible URLs that we generate will not result in a 404. You could base it off of this script (but should probably parallelise the requests with the `AsyncClient`)

@RobertCraigie will do it later. Just pushed the basic implementation and now I'm going to sleep. It should be on par with the default TypeScript one. 
Tested it on Windows. Wanted to test on Debian but Prisma is not built for the newest OpenSSL 3.0 😩

**Comment by RobertCraigie** [comment]
@HKGx No worries, thank you!

> Wanted to test on Debian but Prisma is not built for the newest OpenSSL 3.0 😩

I hadn't even realised there was an OpenSSL 3.0 to be honest, is it worth making an issue in https://github.com/prisma/prisma-engines?

**Comment by RobertCraigie** [comment]
@HKGx I've pushed some minor changes and small fixes.

Some thoughts:

- I'm happy with the name change of `PRISMA_CLI_URL` to `PRISMA_CLI_MIRROR` and `PRISMA_ENGINES_MIRROR` but we need to document that somewhere (even if it wasn't documented before) as it is a breaking change.

Regarding code style:

- I prefer single quotes `'a'` over double quotes `"b"`
- I generally think if you need to create a method like: `engine_url_for` then that probably belongs within a class

Great work so far!

File: src/prisma/binaries/platform.py [hunk 1]
```diff
@@ -1,75 +1,125 @@
...
+import sys
 from functools import lru_cache
-from typing import Tuple
-
-
-def name() -> str:
-    return _platform.system().lower()
+from typing import Optional
 
+import distro  # type: ignore # pyright: reportMissingTypeStubs=false
```
Comment (RobertCraigie) [inline src/prisma/binaries/platform.py:8]:
```suggestion
import distro  # type: ignore[import]
```
This is a subtle change but means that both mypy and pyright will not report an error for missing type stubs but will still type check it. Previously pyright would disregard distros typings. You could either make this change wherever distro is imported or add an entry to `_compat.py`:

```py
import distro as distro  # type: ignore[import]
```
I haven't tested this but it should work.

**Pull request closed without merging**